pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1074 @@
+//!
+//! Orders are executed on block initialize and they are sorted based on randomness derived from relay chain block hash.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::MaxEncodedLen;
+
+use cumulus_primitives_core::relay_chain::Hash;
+use frame_support::{
+	ensure,
+	pallet_prelude::*,
+	traits::{Get, Len},
+	transactional,
+	weights::WeightToFee as FrameSupportWeight,
+};
+use frame_system::{ensure_signed, pallet_prelude::OriginFor, Origin};
+use hydradx_traits::pools::SpotPriceProvider;
+use hydradx_traits::router::PoolType;
+use hydradx_traits::{OraclePeriod, PriceOracle};
+use orml_traits::arithmetic::CheckedAdd;
+use orml_traits::MultiCurrency;
+use orml_traits::NamedMultiReservableCurrency;
+use pallet_route_executor::Trade;
+use pallet_route_executor::TradeAmountsCalculator;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use scale_info::TypeInfo;
+use sp_runtime::traits::CheckedMul;
+use sp_runtime::traits::One;
+use sp_runtime::{
+	traits::{BlockNumberProvider, Saturating},
+	ArithmeticError, BoundedVec, DispatchError, FixedPointNumber, FixedU128, Permill,
+};
+use sp_std::vec::Vec;
+use sp_std::{
+	cmp::{max, min},
+	vec,
+};
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod types;
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+use crate::types::*;
+
+type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+
+pub const SHORT_ORACLE_BLOCK_PERIOD: u32 = 10;
+pub const RETRY_TO_SEARCH_FOR_FREE_BLOCK: u32 = 5;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::HasCompact;
+
+	use frame_support::weights::WeightToFee;
+
+	use frame_system::pallet_prelude::OriginFor;
+	use hydra_dx_math::ema::EmaPrice;
+	use hydradx_traits::pools::SpotPriceProvider;
+	use hydradx_traits::PriceOracle;
+	use orml_traits::NamedMultiReservableCurrency;
+	use sp_runtime::DispatchError::BadOrigin;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+	where
+		<T as pallet_route_executor::Config>::AssetId: From<<T as pallet::Config>::Asset>,
+		<T as pallet_route_executor::Config>::Balance: From<u128>,
+		u128: From<<T as pallet_route_executor::Config>::Balance>,
+	{
+		fn on_initialize(current_blocknumber: T::BlockNumber) -> Weight {
+			let mut weight = <T as pallet::Config>::WeightInfo::on_initialize_with_empty_block();
+
+			let Ok(mut random_generator) = T::RandomnessProvider::generator() else {
+				Self::deposit_event(Event::RandomnessGenerationFailed {
+					block: current_blocknumber,
+				});
+				return weight;
+			};
+
+			let mut schedule_ids: Vec<ScheduleId> = ScheduleIdsPerBlock::<T>::get(current_blocknumber).to_vec();
```

Comment:
use `take` as it removes the entry.
```suggestion
			let mut schedule_ids: Vec<ScheduleId> = ScheduleIdsPerBlock::<T>::take(current_blocknumber).to_vec();
```